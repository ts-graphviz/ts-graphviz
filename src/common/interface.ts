export interface IDot {
  toDot(): string;
}
