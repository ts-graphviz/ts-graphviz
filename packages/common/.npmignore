src
tests
tsconfig.json
typedoc.json
vite.config.ts
