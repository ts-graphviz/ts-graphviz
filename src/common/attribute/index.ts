export * from './attribute.js';
export * from './keys.js';
