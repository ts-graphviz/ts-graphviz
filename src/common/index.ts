export * from './type/index.js';
export * from './attribute/index.js';
export * from './models.js';
