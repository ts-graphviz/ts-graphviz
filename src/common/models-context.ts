import {
  GraphAttributesObject,
  RootGraphModel,
  SubgraphAttributesObject,
  SubgraphModel,
  NodeAttributesObject,
  NodeModel,
  EdgeTargetTuple,
  EdgeAttributesObject,
  EdgeModel,
} from './models.js';

/**
 * @beta
 */
export interface RootGraphConstructor {
  new (id?: string, attributes?: GraphAttributesObject): RootGraphModel;
  new (id?: string, strict?: boolean, attributes?: GraphAttributesObject): RootGraphModel;
  new (strict?: boolean, attributes?: GraphAttributesObject): RootGraphModel;
  new (attributes?: GraphAttributesObject): RootGraphModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): RootGraphModel;
}

/**
 * @beta
 */
export interface SubgraphConstructor {
  new (id?: string, attributes?: SubgraphAttributesObject): SubgraphModel;
  new (attributes?: SubgraphAttributesObject): SubgraphModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): SubgraphModel;
}

/**
 * @beta
 */
export interface NodeConstructor {
  new (id: string, attributes?: NodeAttributesObject): NodeModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): NodeModel;
}

/**
 * @beta
 */
export interface EdgeConstructor {
  new (targets: EdgeTargetTuple, attributes?: EdgeAttributesObject): EdgeModel;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  new (...args: any[]): EdgeModel;
}

/**
 * @beta
 */
export interface ModelsContext {
  Graph: RootGraphConstructor;
  Digraph: RootGraphConstructor;
  Subgraph: SubgraphConstructor;
  Node: NodeConstructor;
  Edge: EdgeConstructor;
}

/**
 * @alpha
 */
export const RootModelsContext: ModelsContext = Object.seal({
  Graph: null,
  Digraph: null,
  Subgraph: null,
  Node: null,
  Edge: null,
} as unknown as ModelsContext);

/**
 * @alpha
 */
export function createModelsContext(models: Partial<ModelsContext>): ModelsContext {
  return Object.assign(Object.seal(Object.assign({}, RootModelsContext)), models);
}
