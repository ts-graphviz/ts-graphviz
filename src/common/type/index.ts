/**
 * @module ts-graphviz/type
 */
export * from './types.js';
