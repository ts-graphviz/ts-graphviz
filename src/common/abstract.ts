import { IDot } from '../common';
// tslint:disable: max-classes-per-file

/**
 * @category Base
 */
export abstract class GraphvizObject {}

/**
 * @category Base
 */
export abstract class DotBase extends GraphvizObject implements IDot {
  protected static quoteString(src: string): string {
    return `"${src}"`;
  }

  protected static joinLines(...lines: (string | undefined)[]): string {
    return lines.filter(l => typeof l === 'string').join('\n');
  }

  protected static indent(src: string, depth: number = 1): string {
    const indent = ' '.repeat(2 * depth);
    return src
      .split('\n')
      .map(l => `${indent}${l}`)
      .join('\n');
  }

  public abstract toDot(): string;
}
