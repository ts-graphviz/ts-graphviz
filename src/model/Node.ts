import { DotBase } from '../common';
import { NodeAttributes } from './attributes';

/**
 * @category Primary
 */
export class Node extends DotBase {
  public readonly attributes: NodeAttributes = new NodeAttributes();
  constructor(public readonly id: string) {
    super();
  }

  public toDot(): string {
    const target = Node.quoteString(this.id);
    const attrs = this.attributes.size > 0 ? ` ${this.attributes.toDot()}` : '';
    return `${target}${attrs};`;
  }
}
