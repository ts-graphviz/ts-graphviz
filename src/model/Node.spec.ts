import { DotBase, GraphvizObject } from '../common';
import { Node } from './Node';

describe('class Node', () => {
  let node: Node;
  beforeEach(() => {
    node = new Node('test');
  });

  it('should be instance of Node/DotBase/GraphvizObject', () => {
    expect(node).toBeInstanceOf(Node);
    expect(node).toBeInstanceOf(DotBase);
    expect(node).toBeInstanceOf(GraphvizObject);
  });

  describe('renders correctly by toDot method', () => {
    it('simple node', () => {
      expect(node.toDot()).toMatchSnapshot();
    });

    it('has some attributes', () => {
      node.attributes.set('label', 'this is test');
      node.attributes.set('color', 'red');
      expect(node.toDot()).toMatchSnapshot();
    });
  });
});
