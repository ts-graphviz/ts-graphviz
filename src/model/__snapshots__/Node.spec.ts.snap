// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class Node renders correctly by toDot method has some attributes 1`] = `
"\\"test\\" [
  label = \\"this is test\\",
  color = \\"red\\",
];"
`;

exports[`class Node renders correctly by toDot method simple node 1`] = `"\\"test\\";"`;
